pygmt/src/dimfilter.py

```diff
@@ -0,0 +1,143 @@
+        computed only once. The next two options are slower because they
+        recompute weights for each latitude.
+
+        - **3**\ : grid (x,y) in degrees, *width* in km, dx scaled by
+          cosine(y), Cartesian distance calculation.
+        - **4**\ : grid (x,y) in degrees, *width* in km, Spherical distance
+          calculation.
+    filter : str
+        **x**\ *width*\ [**+l**\|\ **u**]
+        Sets the primary filter type. Choose among convolution and
+        non-convolution filters. Append the filter code **x** followed by
+        the full diameter *width*. Available convolution filters are:
+
+        - (**b**) Boxcar: All weights are equal.
+        - (**c**) Cosine Arch: Weights follow a cosine arch curve.
+        - (**g**) Gaussian: Weights are given by the Gaussian function.
+
+        Non-convolution filters are:
+
+        - (**m**) Median: Returns median value.
+        - (**p**) Maximum likelihood probability (a mode estimator): Return
+          modal value. If more than one mode is found we return their average
+          value. Append **+l** or **+h** to the filter width if you want
+          to return the smallest or largest of each sector's modal values.
+    sectors : str
+        **x**\ *sectors*\ [**+l**\|\ **u**]
+        Sets the secondary filter type **x** and the number of bow-tie sectors.
+        *sectors* must be integer and larger than 0. When *sectors* is
+        set to 1, the secondary filter is not effective. Available secondary
+        filters **x** are:
+
+        - (**l**) Lower: Return the minimum of all filtered values.
+        - (**u**) Upper: Return the maximum of all filtered values.
+        - (**a**) Average: Return the mean of all filtered values.
+        - (**m**) Median: Return the median of all filtered values.
+        - (**p**) Mode: Return the mode of all filtered values:
+          If more than one mode is found we return their average
+          value. Append **+l** or **+h** to the sectors if you rather want to
+          return the smallest or largest of the modal values.
+    spacing : str
+        *x_inc* [and optionally *y_inc*] is the output Increment. Append
+        **m** to indicate minutes, or **c** to indicate seconds. If the new
+        *x_inc*, *y_inc* are NOT integer multiples of the old ones (in the
+        input data), filtering will be considerably slower. [Default: Same
+        as input.]
+    region : str or list
+        [*west*, *east*, *south*, *north*].
+        Defines the region of the output points. [Default: Same as input.]
+    {V}
+
+    Returns
+    -------
+    ret: xarray.DataArray or None
+        Return type depends on whether the ``outgrid`` parameter is set:
+
+        - :class:`xarray.DataArray` if ``outgrid`` is not set
+        - None if ``outgrid`` is set (grid output will be stored in file set by
+          ``outgrid``)
+    """
+    if ("D" not in kwargs) or ("F" not in kwargs) or ("N" not in kwargs):
```

Comment:
https://github.com/GenericMappingTools/pygmt/blob/c306352a0c9f965c0fd52cc4f46f1a6074bb58b4/pygmt/helpers/utils.py#L260

We have the `args_in_kwargs` function to simplify line 128.

Reply:
Unless I misunderstand its function, isn't `args_in_kwargs` to test if _any_ of the arguments are in the kwargs? This is making sure that all 3 arguments are in the kwargs.

Reply:
There must be some misunderstanding here. The exception message says "At least one of the following parameters must be specified: distance, filters, or sectors.", but `if ("D" not in kwargs) or ("F" not in kwargs) or ("N" not in kwargs):` requires all the three are in the `kwargs`.

Reading the dimfilter documentation (https://docs.generic-mapping-tools.org/dev/dimfilter), a command like `gmt dimfilter depths.txt -Q > output.z` is also valid.

Reply:
So the check should probably look something like this:

```suggestion
    if not all(arg in kwargs for arg in ["D", "F", "N"]) and "Q" not in kwargs:
```

However, the `-Q` mode requires a table input, and that has not been properly implemented in this Pull Request yet. Do you think we should just have the `grid`->`grid` functionality finalized, and then do `table`->`grid` in a separate PR?

Reply:
@weiji14 I would prefer to get this PR wrapped up, rather than try and work out table to grid functionality. I'm assuming that means leaving out the mention of `Q` in this if statement?